packages/accounts/src/services/accounts.ts

```diff
@@ -959,4 +933,322 @@ export class AccountsService implements AccountsServiceInterface {
+                }
+                if (!lentBalanceId) {
+                  lentBalanceId = randomId()
+                  newBalanceIds[account.superAccount.id] = {
+                    lentBalanceId
+                  }
+                }
+              } else if (!lentBalanceId) {
+                throw new UnknownBalanceError(account.superAccount.id)
+              }
+              transfers.push({
+                sourceBalanceId: lentBalanceId,
+                destinationBalanceId: borrowedBalanceId,
+                amount
+              })
+              if (!account.superAccount.superAccountId) {
+                transfers.push({
+                  sourceBalanceId: account.superAccount.balanceId,
+                  destinationBalanceId: accountWithSuperAccounts.balanceId,
+                  amount
+                })
+              }
+            } else if (
+              mode === AdjustTrustlineMode.Settle ||
+              mode === AdjustTrustlineMode.Revolve
+            ) {
+              if (!account.borrowedBalanceId) {
+                throw new UnknownBalanceError(account.id)
+              } else if (!account.superAccount.lentBalanceId) {
+                throw new UnknownBalanceError(account.superAccount.id)
+              }
+              transfers.push({
+                sourceBalanceId: account.borrowedBalanceId,
+                destinationBalanceId: account.superAccount.lentBalanceId,
+                amount
+              })
+              if (!account.superAccount.superAccountId) {
+                transfers.push({
+                  sourceBalanceId: accountWithSuperAccounts.balanceId,
+                  destinationBalanceId: account.superAccount.balanceId,
+                  amount
+                })
+              }
+            }
+          }
+
+          const newBalances: BalanceOptions[] = []
+          for (const account in newBalanceIds) {
+            for (const balance in newBalanceIds[account]) {
+              newBalances.push({
+                id: newBalanceIds[account][balance],
+                flags:
+                  0 |
+                  (['trustlineBalanceId', 'borrowedBalanceId'].includes(balance)
+                    ? AccountFlags.debits_must_not_exceed_credits
+                    : AccountFlags.credits_must_not_exceed_debits) |
+                  AccountFlags.linked
+              })
+            }
+          }
+          if (newBalances.length) {
+            newBalances[newBalances.length - 1].flags &= ~AccountFlags.linked
+            await this.createBalances(
+              newBalances,
+              accountWithSuperAccounts.assetScale
+            )
+            for (const accountId in newBalanceIds) {
+              await IlpAccountModel.query()
+                .patch(newBalanceIds[accountId])
+                .findById(accountId)
+            }
+          }
+        }
+      )
+      if (err) {
+        return err
+      }
+      const res = await this.client.createTransfers(
```

Comment:
How does (or can) things recover if the postgres transaction above succeeds but the tigerbeetle transaction fails (e.g. network trouble)?

Reply:
In that case, the ids of any new TigerBeetle accounts that were created by `adjustTrustline` will remain in postgres. Those TigerBeetle accounts will be used for any subsequent relevant trustline transfers for the Interledger account.